### PR TITLE
New Hashset

### PR DESCRIPTION
**Pull request by oarap**
This PR provides better implementation of `CHashSet`. The improvements follows the same logic as `CHashMap` and provided a new iterator called `CHashSetIter`. Even though `CHashSet` exists in the code base, there was no real usage in the Orca codebase. 

I have provided initially two use cases in this PR:
1. Deduplication of `CExpression`s
2. Column statistics configuration

We were unnecessarily allocating and deallocating <key,value> pairs while we only need <elements> for the logic of both cases provided.

### CODE REVIEW

File: libgpos/server/src/unittest/gpos/common/CHashSetIterTest.cpp [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+		
+		// iterate over full set
+		SetIter si(ps);
+		while (si.FAdvance())
+		{
+			ulpChkSumElement += *(si.Pt());
+		}
+		
+		// use Gauss's formula for checksum-ing
+		GPOS_ASSERT(ulpChkSumElement == ((ul + 2) * (ul + 1)) / 2);
```
Comment (khannaekta) [inline libgpos/server/src/unittest/gpos/common/CHashSetIterTest.cpp]: While checking the sum one by one is mathematically correct, this is a bit on the obscure side. Can we make the test more straightforward by just loading all elements through the iterator into an array or dynamic array, sort it, then assert its exact content?

jesse and ekta

File: libnaucrates/include/naucrates/md/IMDId.h [hunk 2]
```diff
@@ -219,12 +219,10 @@ namespace gpmd
 	typedef CDynamicPtrArray<IMDId, CleanupRelease> DrgPmdid;
 
     // hash map from mdid to mdid
```
Comment (khannaekta) [inline libnaucrates/include/naucrates/md/IMDId.h]: hash map -> hash set

File: libnaucrates/include/naucrates/md/IMDId.h [hunk 3]
```diff
@@ -219,12 +219,10 @@ namespace gpmd
...
-    typedef CHashMap<IMDId, IMDId, IMDId::UlHashMDId, IMDId::FEqualMDId,
-    		CleanupRelease<IMDId>, CleanupRelease<IMDId> > HMMDIdMDId;
+    typedef CHashSet<IMDId, IMDId::UlHashMDId, IMDId::FEqualMDId, CleanupRelease<IMDId> > HSMDId;
 
     // iterator over the hash map from column id to information of missing statistics column info
```
Comment (khannaekta) [inline libnaucrates/include/naucrates/md/IMDId.h]: ditto

File: libgpos/include/gpos/common/CHashSet.h [hunk 4]
```diff
@@ -101,6 +110,13 @@ namespace gpos
 			typedef CDynamicPtrArray<CHashSetElem, CleanupDelete> DrgHashChain;
 			DrgHashChain **m_ppdrgchain;
 
+			// array for elements
```
Comment (vraghavan78) [inline libgpos/include/gpos/common/CHashSet.h]: array of elements

File: libgpos/include/gpos/common/CHashSet.h [hunk 5]
```diff
@@ -101,6 +110,13 @@ namespace gpos
...
+			// We use CleanupNULL because the elements are owned by the hash table
+			typedef CDynamicPtrArray<T, CleanupNULL> DrgElements;
```
Comment (vraghavan78) [inline libgpos/include/gpos/common/CHashSet.h]: CleanupNULL is a NOOP. I am confused by the comment.

File: libgpos/include/gpos/common/CHashSet.inl [hunk 6]
```diff
@@ -1,20 +1,8 @@
-//---------------------------------------------------------------------------
 //	Greenplum Database
```
Comment (vraghavan78) [inline libgpos/include/gpos/common/CHashSet.inl]: can we move this *.inl into *.cpp?

File: libgpos/server/src/unittest/gpos/common/CHashSetIterTest.cpp [hunk 7]
```diff
@@ -0,0 +1,81 @@
+//---------------------------------------------------------------------------
+//	Greenplum Database
+//	Copyright (C) 2017 Pivotal Software, Inc
+
```
Comment (vraghavan78) [inline libgpos/server/src/unittest/gpos/common/CHashSetIterTest.cpp]: header seems chopped.

**Pull request merged**